frontend/model/contracts/chatroom.js

```diff
@@ -287,6 +287,7 @@ sbp('chelonia/defineContract', {
         state.members[memberID].leftHeight = height
         delete state.members[memberID]
 
+        if (!state.attributes) return
```

Comment:
These changes in this file is to avoid errors that happens when the group private chatroom is being synced by the group members who are not part of this private chatroom.

Reply:
I don't think `state.members` would be `undefined` here, because `state.members` is already initiated inside `/join/process` when the contract is synced by someone who is not part of. But since the `attributes` is not initiated in any places, actually it was making errors and also notifications of type `CHELONIA_ERROR`.

In fact, this code and workflow are what @corrideat made and I think it's to handle the following situation.
- [Tab1] user1 creates a group
- [Tab2] user2 joins the group
- [Tab3] user3 joins the group
- [Tab2] user2 creates a private chatroom and invites user3
- [Tab2] user2 creates a proposal to remove user3
- [Tab1] user1 approves the proposal and kicks user3 from the group (and also kicks user3 from all the group chatrooms including the private ones)
In this case, the actorID is the user1 and he doesn't have the signing key of private chatroom. In this case, I can see that the user1 users the group contract key to kick user3 from the private chatroom.
In this case, the chatroom constructor can not be synced and user1 can only sync the actions from when user2 joins the group. So `state.members`, `state.attributes`, `state.messages`, ... are all `undefined` and we should make error handlers manually.

I am not sure I have 100% understood his code, but above is based on my understanding.

Reply:
>  So state.members, state.attributes, state.messages, ... are all undefined and we should make error handlers manually.

You just said it yourself, if `state.members` is undefined then the code above online 287 would throw as well.

Reply:
`state.members` can only be `undefined` in `/join/process`. And we manually initiate it in the line 178.
```
'gi.contracts/chatroom/join': {
  validate: actionRequireInnerSignature(objectOf({
    memberID: optional(string) // user id of joining memberID
  })),
  process ({ data, meta, hash, height, contractID, innerSigningContractID }, { state }) {
    const memberID = data.memberID || innerSigningContractID

    if (!memberID) {
      throw new Error('The new member must be given either explicitly or implcitly with an inner signature')
    }

    if (!state.renderingContext) {
      if (!state.members) {
        state['members'] = {}
      }
      if (state.members[memberID]) {
        throw new GIChatroomAlreadyMemberError(`Can not join the chatroom which ${memberID} is already part of`)
      }
    }

...
```

Reply:
I tested your theory above while removing these checks and did get an error:

```
[chelonia] ERROR 'TypeError' in processMutation for <op_a|z9brRu3VK2pPouEujA3sxDkiRsb1Zz8MsGaxw222wgeTxTe8CoT5 of z9brRu3VViuiGEcbSsroBWZ3GiqNUjNxvaRmaCye5R62q1sx15CZ>: can't access property "type", state.attributes is undefined TypeError: can't access property "type", state.attributes is undefined
```

So it seems like you're right.

Could you please copy/paste the steps you mentioned up above and add a comment with them to the code so that we remember why this is being done?

Also, it seems one check is still missing. After following the steps above I got this error:

```
[gi.contracts/chatroom/leave/sideEffect] Error at sideEffect can't access property "privacyLevel", state.attributes is undefined
```

![Screenshot 2024-07-17 at 7 38 21 PM](https://github.com/user-attachments/assets/17bc34bb-1077-479b-b53a-7316db974814)